library/core/src/intrinsics.rs

```diff
@@ -2138,6 +2138,30 @@ pub const unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: us
     }
 }
 
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub struct Invariant {
+    offset: u64,
+    size: u64,
+    start: u128,
+    end: u128,
+}
```

Comment:
What does an `Invariant` *mean*? As in, what is the actual invariant that is encoded by an element of `Invariant`?

Reply:
Each Invariant in the list means "read off an integer with byte size `size`, `offset` bytes past the start of the struct, and then assert that it is inside the WrappingRange { start, end }. (Where WrappingRange is https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/struct.WrappingRange.html )

So for `(NonZeroU16, u8, bool)`, the list will be

* `offset: 0, size: 2, start: 1, end: 65535`
* `offset: 2, size: 1, start: 0, end: 255`
* `offset: 3, size: 1, start: 0, end: 1`.
(assuming that's laid out in order)

I'll change `size` to be an enum so the match on it is exhaustive, and so I can add "is a pointer" so we're not doing ptr to int transmutes.

We still generate an Invariant for types like `u8` so that we branch on their value, so memory sanitizer picks it up.

Though we could only enable generating invariants for types that are fully valid at any bit-pattern if memory sanitizer is enabled, since that's the only time they're useful.

Reply:
Ah, I see. I think `valid_range_{start,end}` would be better names for those variables then. Please also add a doc comment explaining the meaning of the fields.